Here is how I verified this fix.

Starting with empty build cache (rm -r ~/.gradle/caches/build-cache-1), I introduced a test that always fails and ran the following gradlew command.

```
./gradlew --info --build-cache --no-scan --continue -PignoreFailures=true :metadata:test
```

The tests all run, Gradle exits with return code 1, and the following output is produced:

```
> Task :metadata:test FAILED

933 tests completed, 1 failed, 1 skipped
Finished generating test XML results (0.036 secs) into: /Users/davidarthur/Code/Apache/kafka/metadata/build/test-results/test
Generating HTML test report...
Finished generating test html results (0.043 secs) into: /Users/davidarthur/Code/Apache/kafka/metadata/build/reports/tests/test
There were failing tests. See the report at: file:///Users/davidarthur/Code/Apache/kafka/metadata/build/reports/tests/test/index.html
Copy JUnit XML for metadata to /Users/davidarthur/Code/Apache/kafka/build/junit-xml/metadata
[ant:copy] Copying 105 files to /Users/davidarthur/Code/Apache/kafka/build/junit-xml/metadata

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/davidarthur/Code/Apache/kafka/build.gradle' line: 528

* What went wrong:
Execution failed for task ':metadata:test'.
> Failing this task since 'metadata:test' had test failures.

BUILD FAILED in 14s
31 actionable tasks: 7 executed, 24 up-to-date
```

Notably, we see the JUnit XML files moved to the right place and the task ending with failure.

If run again with no changes, the outcome is the same. After removing the failing test and re-running the gradlew command, I see the following:

```
> Task :metadata:test
Finished generating test XML results (0.046 secs) into: /Users/davidarthur/Code/Apache/kafka/metadata/build/test-results/test
Generating HTML test report...
Finished generating test html results (0.03 secs) into: /Users/davidarthur/Code/Apache/kafka/metadata/build/reports/tests/test
Copy JUnit XML for metadata to /Users/davidarthur/Code/Apache/kafka/build/junit-xml/metadata
[ant:copy] Copying 105 files to /Users/davidarthur/Code/Apache/kafka/build/junit-xml/metadata
Stored cache entry for task ':metadata:test' with cache key b2da6d0b8f4fbf99a090c0287dbf2f2c

BUILD SUCCESSFUL in 18s
31 actionable tasks: 6 executed, 25 up-to-date
```

The tests are all run, Gradle exits with 0, and we see the cache get populated. Now re-running the same gradlew command results in

```
> Task :metadata:test UP-TO-DATE
Custom actions are attached to task ':metadata:test'.
Build cache key for task ':metadata:test' is b2da6d0b8f4fbf99a090c0287dbf2f2c
Skipping task ':metadata:test' as it is up-to-date.

BUILD SUCCESSFUL in 810ms
31 actionable tasks: 31 up-to-date
```

No XML files moved, exit code 0.

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -108,6 +108,7 @@ jobs:
         # --continue:     Keep running even if a test fails
         # -PcommitId      Prevent the Git SHA being written into the jar files (which breaks caching)
         timeout-minutes: 180  # 3 hours
+        continue-on-error: true
         run: |
           ./gradlew --build-cache --scan --continue \
           -PtestLoggingEvents=started,passed,skipped,failed \

--- a/build.gradle
+++ b/build.gradle
@@ -471,10 +471,12 @@ subprojects {
 
   test {
     ext {
+      isGithubActions = System.getenv('GITHUB_ACTIONS') != null
       hadFailure = false  // Used to track if any tests failed, see afterSuite below
     }
+
     maxParallelForks = maxTestForks
-    ignoreFailures = true
+    ignoreFailures = userIgnoreFailures || ext.isGithubActions
 
     maxHeapSize = defaultMaxHeapSize
     jvmArgs = defaultJvmArgs
@@ -513,19 +515,19 @@ subprojects {
     // This closure will copy JUnit XML files out of the sub-project's build directory and into
     // a top-level build/junit-xml directory. This is necessary to avoid reporting on tests which
     // were not run, but instead were restored via FROM-CACHE. See KAFKA-17479 for more details.
-    //
-    // This closure will not run if a test fails and ignoreFailures is not true.
     doLast {
-      def dest = rootProject.layout.buildDirectory.dir("junit-xml/${project.name}").get().asFile
-      println "Copy JUnit XML for ${project.name} to $dest"
-      ant.copy(todir: "$dest") {
-        ant.fileset(dir: "${test.reports.junitXml.entryPoint}")
-      }
+      if (ext.isGithubActions) {
+        def dest = rootProject.layout.buildDirectory.dir("junit-xml/${project.name}").get().asFile
+        println "Copy JUnit XML for ${project.name} to $dest"
+        ant.copy(todir: "$dest") {
+          ant.fileset(dir: "${test.reports.junitXml.entryPoint}")
+        }
 
-      // If there were any test failures, we want to fail the task to prevent the failures
-      // from being cached.
-      if (ext.hadFailure && !userIgnoreFailures) {
-        throw new GradleException("Failing this task since '${project.name}:${name}' had test failures.")
+        // If there were any test failures, we want to fail the task to prevent the failures
+        // from being cached.
+        if (ext.hadFailure) {
+          throw new GradleException("Failing this task since '${project.name}:${name}' had test failures.")
+        }
       }
     }
   }

--- a/build.gradle
+++ b/build.gradle
@@ -80,7 +80,7 @@ ext {
   maxTestForks = project.hasProperty('maxParallelForks') ? maxParallelForks.toInteger() : Runtime.runtime.availableProcessors()
   maxScalacThreads = project.hasProperty('maxScalacThreads') ? maxScalacThreads.toInteger() :
       Math.min(Runtime.runtime.availableProcessors(), 8)
-  userIgnoreFailures = project.hasProperty('ignoreFailures') ? ignoreFailures : false
+  userIgnoreFailures = project.hasProperty('ignoreFailures') ? ignoreFailures.toBoolean() : false
 
   userMaxTestRetries = project.hasProperty('maxTestRetries') ? maxTestRetries.toInteger() : 0
   userMaxTestRetryFailures = project.hasProperty('maxTestRetryFailures') ? maxTestRetryFailures.toInteger() : 0
@@ -524,7 +524,7 @@ subprojects {
 
       // If there were any test failures, we want to fail the task to prevent the failures
       // from being cached.
-      if (ext.hadFailure) {
+      if (ext.hadFailure && !userIgnoreFailures) {
         throw new GradleException("Failing this task since '${project.name}:${name}' had test failures.")
       }
     }

